File: python/81-90/Largest Rectangle in Histogram.py
    # @param {integer[]} height
    # @return {integer}
    def largestRectangleArea(self, height):
        if len(height) ==0:
            return 0
        height.append(0)
        stack = []
        mx = 0
        for i in xrange(len(height)):
            while len(stack) >0 and height[stack[-1]]>height[i]:
                h = height[stack[-1]]
                del stack[-1]
                left = -1
                if len(stack)>0:
                    left = stack[-1]
                if (h*(i-1-left)>mx):
                    mx = h*(i-1-left)
            stack.append(i)
        return mx

File: Cpp/81-90/Largest Rectangle in Histogram.cpp
/*
    runtime:24ms
    error: 1  
    h = myStack.top() should be h = height[myStack.top()];

*/
class Solution {
public:
    int largestRectangleArea(vector<int>& height) {
        stack<int> myStack;
        if(height.size()==0) return 0;
        height.push_back(0);
        int Max=0;
        for(int i=0; i<height.size(); i++){
            while(!myStack.empty()&&height[i]<height[myStack.top()]){
                int h=height[myStack.top()];
                myStack.pop();
                int wid = myStack.empty()?-1:myStack.top();
                int Area = (i-1-wid)*h;
                if(Area>Max) Max = Area;
            }
            myStack.push(i);
        }
        return Max;
    }
};
